packages/framework/aqueduct/src/data-object-factories/dataObjectFactory.ts

```diff
@@ -55,6 +70,85 @@ export class DataObjectFactory<
 			mergedObjects.push(SharedMap.getFactory());
 		}
 
-		super(type, ctor, mergedObjects, optionalProviders, registryEntries, runtimeFactory);
+		let converted = false;
+		const convertRoundTripP = new Deferred<void>();
+
+		const fullConvertDataFn =
+			convertDataFn === undefined
+				? undefined
+				: async (runtime: FluidDataStoreRuntime) => {
+						if (!converted) {
+							converted = true;
+							submitConversionOp(runtime);
+							await convertRoundTripP.promise;
+
+							const root = (await runtime.getChannel(
+								dataObjectRootDirectoryId,
+							)) as ISharedDirectory;
+							await convertDataFn(runtime, root);
+						}
+					};
+
+		const submitConversionOp = (runtime: FluidDataStoreRuntime): void => {
+			runtime.submitMessage(DataStoreMessageType.ChannelOp, "conversion", undefined);
+		};
+
+		super(
+			type,
+			ctor,
+			mergedObjects,
+			optionalProviders,
+			registryEntries,
+			class ConverterDataStoreRuntime extends runtimeFactory {
```

Comment:
Is it possible to extract this class into top-level class, and pass whatever it needs to constructor (as opposed to using implicit capture of state)?